fix: e-invoice vertical tax calculation

En0Ma1259 opened the pull request:
### 1. Why is this change necessary?
Currently, no difference between `vertical` and `horizontal` tax calculation is made by creating an e-invoice.
So there is a possibility, that small rounding differences occur.

### 2. What does this change do, exactly?
At the moment, there is no way to know which calculation type was used for an order.
Added a new field to the order `taxCalculationType`.
If the order is calculated vertical and is gross, then the tax will be calculated once by `type` and `taxRate`.

Adds payment mean and buyer reference

### 3. Describe each step to reproduce the issue or behaviour.
Create an order while the `salesChannel` has vertical tax calculation selected.
With multiple `lineItems` it can lead into a difference for e.g. `LineTotalAmount` (see [issue](https://github.com/shopware/shopware/issues/9310) description)

### 4. Please link to the relevant issues (if any).
- fixes https://github.com/shopware/shopware/issues/9310
- fixes https://github.com/shopware/shopware/issues/9453

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

En0Ma1259 committed:
fix: e-invoice vertical tax calculation

En0Ma1259 committed:
Changelog

En0Ma1259 committed:
Migration test

En0Ma1259 committed:
Fixed tests

En0Ma1259 committed:
Api aware field

En0Ma1259 committed:
bc

mitelg commented:
I only had a rough look now, but shouldn't the order be already created with the correct tax calculation? why do we need to calculate something new to just generate some document? :thinking: that feels wrong. do we also calculate everything new on PDFs?

En0Ma1259 commented:
> I only had a rough look now, but shouldn't the order be already created with the correct tax calculation? why do we need to calculate something new to just generate some document? 🤔 that feels wrong. do we also calculate everything new on PDFs?

@mitelg On the first look, this is a valid question.
In short: The problem is, we don't have the values as separated, as we need them. We need the calculated taxes for "products" (`lineTotalAmount`), `shippingCosts` - or other upcoming costs (`chargeAmount`) and `promotions` - everything that reduces the price (`allowanceAmount`).
With the current structure, we are not able to do this. E.g. products are in the same spot as promotions and that alone will not work. So we need to calculate on our own for e-invoices
And we will currently not know which calculation type we should follow.

And no. For PDF's we don't need this, because we will not show the calculated tax for each product. Therefore, this is not a problem

mitelg commented:
@En0Ma1259 I see, thanks for the explanation :+1: 

would it be possible to reuse the class, that is alone responsible for calculating the price? :thinking: `\Shopware\Core\Checkout\Cart\Price\AmountCalculator`

cyl3x commented on src/Core/Migration/V6_6/Migration1747746986OrderTaxCalculationType.php:
```diff
@@ -0,0 +1,24 @@
+
+    public function update(Connection $connection): void
+    {
+        $this->addColumn($connection, 'order', 'tax_calculation_type', 'varchar(255)');
```
Length can be reduced to be the same as in [`sales_channel`](https://github.com/shopware/shopware/blob/636d59ea457b5cdf77388760830cb3638c79001c/src/Core/Migration/V6_3/Migration1589178550AddTaxCalculationType.php#L24):
```suggestion
        $this->addColumn($connection, 'order', 'tax_calculation_type', 'varchar(50)');
```

mstegmeyer commented on src/Core/Checkout/Document/Zugferd/ZugferdDocument.php:
```diff
@@ -225,6 +236,38 @@ public function withTaxes(CartPrice $price): self
+        }
+
+        $cashRounding = $order->getItemRounding();
+        if ($cashRounding) {
+            $lineTotalAmount = (new CashRounding())->cashRound($lineTotalAmount, $cashRounding);
+            $chargeAmount = (new CashRounding())->cashRound($chargeAmount, $cashRounding);
+            $allowanceAmount = (new CashRounding())->cashRound($allowanceAmount, $cashRounding);
+        }
```
`CashRounding` is a service. Can we inject it instead of 3 instantiations?

En0Ma1259 replied:
If I can integrate the `\Shopware\Core\Checkout\Cart\Price\AmountCalculator` this will be removed

https://github.com/shopware/shopware/pull/9734#issuecomment-2908723012

En0Ma1259 commented:
@mitelg We don't have a `salesChannelContext` while generating a document. With the new order field, all information required for the `AmountCalculator` is stored in the order. So a `salesChannelContext` is not really needed, but `OrderEntity`.

----
Without creating a dummy context, we maybe could do something like `SalesChannelContext|OrderEntity` as the function parameter for `AmountCalculator::calculate`. The "only" issue would be the difference between `context->getTaxState()` and `order->getTaxStatus`. (FYI: @shopware/product-cc-checkout )

mitelg commented:
hey @En0Ma1259 

I had another thought on that here over the weekend: would it be possible to not calculate something at all while creating documents? Instead every information needed should be persisted once the order is calculated. so during the creation of the order or if an item is removed or added via admin. 

we had several places for calculation of cart/order in Shopware 5 like Storefront, admin module, document generation and they all behaved slightly different, resulting in hundreds of tickets because of missing cents or what not. and now I am afraid, that we are heading in similar direction now here :grimacing: (the issue description of missing cents already rang some alarm bells :sweat_smile:) 

So the order must be calculated only once and in one single place and nowhere else. If you need more information, we should store them to the order itself or to the respective child entities.

WDYT?

En0Ma1259 committed:
Use AmountCalculator

En0Ma1259 committed:
Feature flag

En0Ma1259 committed:
Feature check

En0Ma1259 commented:
On the one hand I would agree, on the other hand we are creating a structure/workflow for a specific use case (e-invoice ZUGFeRD with CII (Cross Industry Invoice) standard).

Shopware has two buckets (shipping costs and line items) and calculates these. For ZUGFeRD we would need to split things into 3 buckets to get `LineTotalAmount`, `ChargeTotalAmount` and `AllowanceTotalAmount`. I do not know what the structure of the UBL standard for electronic invoices looks like. **If** we want to support UBL, maybe we need a further/different structure for the order. Should that be the goal?

----
The recalculation is needed for `vertical` only. Maybe we should discuss it in more detail rather than here, shouldn't we?

mitelg commented:
> Maybe we should discuss it in more detail rather than here, shouldn't we?

yes :+1: count me in :wink:

En0Ma1259 committed:
abs

En0Ma1259 committed:
Merge branch 'trunk' into fix/e-invoice-vertical-calculation

En0Ma1259 committed:
multiplier to calculate tax | diff PHP <8.3 -> 8.4

En0Ma1259 committed:
multiplier to calculate tax | diff PHP <8.3 -> 8.4

En0Ma1259 committed:
Merge branch 'trunk' into fix/e-invoice-vertical-calculation

MalteJanz commented:
~~This might be a related: looks like we have a similar issue with cancellation invoices~~, see:
https://github.com/shopware/shopware/issues/9924

I guess not really related except that it appears to be two different calculations with different tax values 🙈

En0Ma1259 committed:
Merge branch 'trunk' into fix/e-invoice-vertical-calculation

En0Ma1259 committed:
Added payment mapping + bugfixes

mitelg commented on src/Core/Checkout/Document/Zugferd/ZugferdBuilder.php:
```diff
@@ -89,4 +105,28 @@ protected function matchByType(ZugferdDocument $document, OrderLineItemEntity $l
         $this->eventDispatcher->dispatch(new ZugferdInvoiceItemAddedEvent($document, $lineItem, $parentPosition), 'zugferd-item-added.' . $lineItem->getType());
     }
+
+    protected function addPaymentInfo(ZugferdDocument $document, DocumentConfiguration $config, PaymentMethodEntity $paymentMethod): void
```
does this needs to be protected? think about what we said in the meeting, about adding new API :grin:

mitelg commented on src/Core/Checkout/Document/Zugferd/ZugferdDocument.php:
```diff
@@ -250,11 +323,64 @@ protected function getPrice(CalculatedTax $tax): float
         };
     }
+
+    protected function calculateTaxes(string $type, OrderEntity $order, AmountCalculator $calculator): float
```
also here in this class. please revisit again, if they all need to be protected or public

En0Ma1259 committed:
visibility

En0Ma1259 committed:
visibility

En0Ma1259 committed:
more fixes

En0Ma1259 commented:
@MalteJanz @mstegmeyer @cyl3x @mitelg 
With my latest commits, I have fixed some bugs that I have noticed. Also, I have added payment information for our default payment methods and added a buyer reference (order number)

Can you look over it again pls

En0Ma1259 committed:
more fixes

MalteJanz commented on src/Core/Checkout/Document/Zugferd/ZugferdBuilder.php:
```diff
@@ -58,10 +60,23 @@ public function buildDocument(OrderEntity $order, DocumentConfiguration $config,
 
         $this->addLineItems($document, $order->getLineItems());
 
+        $transaction = $order->getTransactions()?->last();
+        if ($transaction !== null) {
+            if ($transaction->getStateMachineState()?->getTechnicalName() === 'paid') {
+                $document->withPaidAmount($order->getAmountTotal());
```
Do we support more than one transaction per order? In that case this should check that all of them are marked as "paid" right? And adding payment methods should include all of them if documents support that 🤔

En0Ma1259 replied:
No. Older transactions are set to cancelled, if a new one appears. So there is only one transaction with `paid`.

Technically, there are states for `paid (partially)`. But we have no such payment methods, so each plugin with such a state, must handle it themselves.

We have something we should look at, if this issue is resolved https://github.com/shopware/shopware/pull/8418

MalteJanz commented on src/Core/Checkout/Document/Zugferd/ZugferdBuilder.php:
```diff
@@ -91,4 +106,21 @@ protected function matchByType(ZugferdDocument $document, OrderLineItemEntity $l
+
+    private function addPaymentInfo(ZugferdDocument $document, DocumentConfiguration $config, PaymentMethodEntity $paymentMethod): void
+    {
+        if ($paymentMethod->getTechnicalName() === 'payment_cashpayment') {
+            $document->getBuilder()->addDocumentPaymentMean(
+                typeCode: (string) ZugferdPaymentMeans::UNTDID_4461_10->value,
+                information: $paymentMethod->getName()
+            );
+        } elseif ($paymentMethod->getTechnicalName() === 'payment_invoicepayment' || $paymentMethod->getTechnicalName() === 'payment_prepayment') {
```
This looks like you likely wanna extend this if your extension provides a new payment method, e.g. what about PayPal? Are we fine with only supporting these methods for the time being with no easy way for extension? 🤔

En0Ma1259 replied:
PayPal (and other plugins) need to listen to `\Shopware\Core\Checkout\Document\Zugferd\ZugferdInvoiceGeneratedEvent::class` and add the payment means by themselves. We can only do that for "our" (Shopware core) payment methods.

PayPal will be another ticket

MalteJanz commented on src/Core/Checkout/Document/Zugferd/ZugferdDocument.php:
```diff
@@ -204,13 +204,14 @@ public function withDiscountItem(OrderLineItemEntity $lineItem): self
 
             if (!Feature::isActive('v6.8.0.0')) {
-                $this->addAllowanceAmount($actualAmount);
+                $actualAmount >= 0 ? $this->addChargeAmount($actualAmount) : $this->addAllowanceAmount($actualAmount);
```
Just a small nitpick: I would prefer an real if statement here because you have no assignment of the return values and calling different code based on a condition

En0Ma1259 committed:
real if

MalteJanz commented on src/Core/Checkout/Document/Zugferd/ZugferdDocument.php:
```diff
@@ -204,19 +204,24 @@ public function withDiscountItem(OrderLineItemEntity $lineItem): self
 
             if (!Feature::isActive('v6.8.0.0')) {
-                $actualAmount >= 0 ? $this->addChargeAmount($actualAmount) : $this->addAllowanceAmount($actualAmount);
+                if ($isCharge) {
```
This condition is now different than before the change to a real if condition, e.g. not depending on the loop variable `$actualAmount` anymore. Was this intentional and is this the correct condition?

En0Ma1259 replied:
Yes, it was intentional. Overall it is the same anyway. So we have just one check instead of three. The price can't be positive and the tax negative.

MalteJanz commented on src/Core/Checkout/Document/Zugferd/ZugferdDocument.php:
```diff
@@ -204,19 +204,24 @@ public function withDiscountItem(OrderLineItemEntity $lineItem): self
                 ...[
                     'actualAmount' => abs($actualAmount),
-                    'isCharge' => $actualAmount >= 0,
+                    'isCharge' => $isCharge,
```
see above

ghost commented:
Will there be a way to add custom XML data, such as <ram:SpecifiedTradePaymentTerms>?
Sincethe `zugpferdBuilder` is protected, read-only, and not accessible from the event.

En0Ma1259 commented:
> Will there be a way to add custom XML data, such as ram:SpecifiedTradePaymentTerms? Sincethe `zugpferdBuilder` is protected, read-only, and not accessible from the event.

Hi, I have added a new public method to access the builder
https://github.com/shopware/shopware/pull/9734/files#diff-e817d43ca185e0a31bcf7b26434f08df7da0503838f63030df2a1ce3c3e1caf0R294
